src/libopensc/pkcs15-pin.c

```diff
@@ -367,15 +367,19 @@ _sc_pkcs15_verify_pin(struct sc_pkcs15_card *p15card, struct sc_pkcs15_object *p
 		data.pin_reference = skey_info->key_reference;
 	}
 
-	if(p15card->card->reader->capabilities & SC_READER_CAP_PIN_PAD) {
-		if (!pincode && !pinlen)
-			data.flags |= SC_PIN_CMD_USE_PINPAD;
+	if(p15card->card->reader->capabilities & SC_READER_CAP_PIN_PAD && !pincode && !pinlen) {
+		data.flags |= SC_PIN_CMD_USE_PINPAD;
 
 		if (auth_info->attrs.pin.flags & SC_PKCS15_PIN_FLAG_SO_PIN)
 			data.pin1.prompt = "Please enter SO PIN";
 		else
 			data.pin1.prompt = "Please enter PIN";
 	}
+	else {
+		r = _validate_pin(p15card, auth_info, pinlen);
```

Comment:
To be consistent with the existing PIN checking in `sc_pkcs15_change_pin` and `sc_pkcs15_unblock_pin`, this call belongs into `sc_pkcs15_verify_pin`